### DISCUSSION TITLE
Using hook in Higher Order Component (NextJs)

### DISCUSSION THREAD

**Discussion by qtv100291** [eslint/eslint#15161] [Q&A]
I have a Component like below:

```
import React from "react";
import { useRouter } from "next/router";
import authService from "../../services/loginService";

const ProtectedRoute = (WrappedComponent) => (props) => {
    const router = useRouter();
    if (typeof window !== "undefined") {
      const tokenKey = authService.checkToken();
      const timeNow = Date.now() / 1000;
      if (!tokenKey || tokenKey.exp <= timeNow) router.replace('/dang-nhap')
      return <WrappedComponent {...props}/>
    }
    return null
  };

export default ProtectedRoute;
```

It work well in dev mode, but when building production, an eslint error appears 
"React Hooks Error: Hooks can only be called at the top level in the body of a function component"
So I change the code like below :

```
import React from "react";
import { useRouter } from "next/router";
import authService from "../../services/loginService";

const ProtectedRoute = (WrappedComponent) =>{
 const router = useRouter();
 return (props) => {
    if (typeof window !== "undefined") {
      const tokenKey = authService.checkToken();
      const timeNow = Date.now() / 1000;
      if (!tokenKey || tokenKey.exp <= timeNow) router.replace('/dang-nhap')
      return <WrappedComponent {...props}/>
    }
    return null
  }};

export default ProtectedRoute;
```

 Eslint doesn't display error anymore, but in dev mode, nextjs return error : 
"Invalid hook call. Hooks can only be called inside of the body of a function component 
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app"

This problem make me so confused and I dont know how to fix this for working with both eslint and nextjs 
Anyone have any ideas ? Thanks.

**Comment by nzakas** [comment]
You’ll need to talk to the NextJS folks for help as this is related to a rule you are using and not ESLint in general.
